cmd/list.go

```diff
@@ -0,0 +1,280 @@
+	// Create a PathExp based on flags. This is the search space that
+	// will be used to retrieve credentials.
+	filterPathExp, err := pathexp.New(
+		orgName,
+		projectName,
+		envFilters,
+		serviceFilters,
+		instanceFilters,
+		idenityFilters)
+	if err != nil {
+		return errs.NewErrorExitError("Failed to create path for specified flags.", err)
+	}
+
+	// Retrieve envs, services and credentials in parallel
+	var getEnvsServicesCreds sync.WaitGroup
+	getEnvsServicesCreds.Add(3)
+
+	var environments []envelope.Environment
+	var services []envelope.Service
+	var credentials []apitypes.CredentialEnvelope
+	var eErr, sErr, cErr error
+
+	go func() {
+		// Get environments
+		environments, eErr = listEnvs(&c, client, org.ID, project.ID, nil)
+		getEnvsServicesCreds.Done()
+	}()
+
+	go func() {
+		// Get services
+		services, sErr = listServices(&c, client, org.ID, project.ID, nil)
+		getEnvsServicesCreds.Done()
+	}()
+
+	go func() {
+		// Get credentials
+		credentials, cErr = client.Credentials.Search(c, filterPathExp.String())
+		getEnvsServicesCreds.Done()
+	}()
+
+	getEnvsServicesCreds.Wait()
+
+	if cErr != nil {
+		return errs.NewErrorExitError("Could not retrieve credentials.", cErr)
+	}
+
+	if eErr != nil {
+		return errs.NewErrorExitError("Could not retrieve environments.", eErr)
+	}
+
+	if sErr != nil {
+		return errs.NewErrorExitError("Could not retrieve services.", sErr)
+	}
+
+	filteredEnvNames := []string{}
+	filteredServiceNames := []string{}
+
+	// Filter out the retrieved environments based on the
+	// search space provided in filterPathExp. If no flags
+	// were set, all environments will pass the following test.
+	for _, e := range environments {
+		if filterPathExp.Envs.Contains(e.Body.Name) {
+			filteredEnvNames = append(filteredEnvNames, e.Body.Name)
+		}
+	}
+
+	// Filter out the retrieved services based on the
+	// search space provided in filterPathExp. If no flags
+	// were set, all services will pass the following test.
+	for _, s := range services {
+		if filterPathExp.Services.Contains(s.Body.Name) {
+			filteredServiceNames = append(filteredServiceNames, s.Body.Name)
+		}
+	}
+
+	// Create credentialsTree
+	// This will be filled in the following section.
+	for _, eName := range filteredEnvNames {
+		credentialsTree[eName] = make(map[string]credentialSet)
+		for _, sName := range filteredServiceNames {
+			credentialsTree[eName][sName] = make(credentialSet)
+		}
+	}
+
+	// Check for each env and service in the filtered list, add
+	// any credentials along that path to that env/service branch
+	// of the credentialsTree
+	projectPath := "/" + orgName + "/" + projectName + "/"
+	for _, e := range filteredEnvNames {
+		for _, s := range filteredServiceNames {
+			builtPathExp, err := pathexp.Parse(projectPath + e + "/" + s + "/*/*")
+			for _, cred := range credentials {
+				if len(args) > 0 && isSecretNameInList((*cred.Body).GetName(), args) == false {
+					continue
+				}
+				credPathExp := (*cred.Body).GetPathExp()
```

Comment:
You're also derferencing the `cred.Body` frequerntly you could just store it as a var then re-use it (e.g. `body :=  *cred.Body`)

Reply:
Yes, always a bit of an internal debate. Nice variable name at the cost of an extra line to assign it, or save the line in a dense spot but with uglier naming. Switched to the added line, probably best.